File: typescript/src/qicore/base/result.ts
/**
 * QiCore v4.0 - Result<T> Monad Implementation
 *
 * Mathematical Contract-Based TypeScript Library
 * Component 1: Result<T> - Monad for error handling (8 operations)
 */

import type { QiError } from "./error.js";

/**
 * Result<T> represents a computation that can either succeed with a value of type T
 * or fail with a QiError. This implements the Result monad pattern for functional
 * error handling without exceptions.
 */
export abstract class Result<T> {
  /**
   * Operation 1: Create successful result
   */
  static success<T>(value: T): Result<T> {
    return new Success(value);
  }

  /**
   * Operation 2: Create failed result
   */
  static failure<T>(error: QiError): Result<T> {
    return new Failure(error);
  }

  /**
   * Operation 3: Map function over success value (Functor law)
   */
  abstract map<U>(fn: (value: T) => U): Result<U>;

  /**
   * Operation 4: FlatMap (bind) - Monadic bind operation
   */
  abstract flatMap<U>(fn: (value: T) => Result<U>): Result<U>;

  /**
   * Operation 5: Map error value
   */
  abstract mapError(fn: (error: QiError) => QiError): Result<T>;

  /**
   * Operation 6: Recover from error with fallback
   */
  abstract recover(fn: (error: QiError) => T): Result<T>;

  /**
   * Operation 7: Unwrap with default value
   */
  abstract unwrapOr(defaultValue: T): T;

  /**
   * Operation 8: Check if result is successful
   */
  abstract isSuccess(): boolean;

  /**
   * Check if result is a failure
   */
  isFailure(): boolean {
    return !this.isSuccess();
  }

  /**
   * Unwrap value (throws if failure)
   */
  abstract unwrap(): T;

  /**
   * Get error (throws if success)
   */
  abstract error(): QiError;

  /**
   * Apply function if success, otherwise return this
   */
  andThen<U>(fn: (value: T) => Result<U>): Result<U> {
    return this.flatMap(fn);
  }

  /**
   * Chain results together
   */
  and<U>(other: Result<U>): Result<U> {
    return this.flatMap(() => other);
  }

  /**
   * Return this if success, otherwise return other
   */
  or(other: Result<T>): Result<T> {
    return this.isSuccess() ? this : other;
  }

  /**
   * Match pattern for handling both success and failure cases
   */
  match<U>(cases: {
    success: (value: T) => U;
    failure: (error: QiError) => U;
  }): U {
    if (this.isSuccess()) {
      return cases.success(this.unwrap());
    }
    return cases.failure(this.error());
  }
}

/**
 * Success case of Result<T>
 */
class Success<T> extends Result<T> {
  constructor(private readonly value: T) {
    super();
  }

  map<U>(fn: (value: T) => U): Result<U> {
    try {
      return Result.success(fn(this.value));
    } catch (error) {
      // If mapping function throws, convert to failure
      return Result.failure({
        category: "RuntimeError",
        message: `Mapping function failed: ${error}`,
        context: { originalValue: this.value, error: String(error) },
        timestamp: Date.now(),
      } as QiError);
    }
  }

  flatMap<U>(fn: (value: T) => Result<U>): Result<U> {
    try {
      return fn(this.value);
    } catch (error) {
      return Result.failure({
        category: "RuntimeError",
        message: `FlatMap function failed: ${error}`,
        context: { originalValue: this.value, error: String(error) },
        timestamp: Date.now(),
      } as QiError);
    }
  }

  mapError(_fn: (error: QiError) => QiError): Result<T> {
    return this; // No error to map
  }

  recover(_fn: (error: QiError) => T): Result<T> {
    return this; // Already successful
  }

  unwrapOr(_defaultValue: T): T {
    return this.value;
  }

  isSuccess(): boolean {
    return true;
  }

  unwrap(): T {
    return this.value;
  }

  error(): never {
    throw new Error("Called error() on Success result");
  }
}

/**
 * Failure case of Result<T>
 */
class Failure<T> extends Result<T> {
  constructor(private readonly errorValue: QiError) {
    super();
  }

  map<U>(_fn: (value: T) => U): Result<U> {
    return new Failure<U>(this.errorValue);
  }

  flatMap<U>(_fn: (value: T) => Result<U>): Result<U> {
    return new Failure<U>(this.errorValue);
  }

  mapError(fn: (error: QiError) => QiError): Result<T> {
    try {
      return Result.failure(fn(this.errorValue));
    } catch (error) {
      // If error mapping function throws, create new error
      return Result.failure({
        category: "RuntimeError",
        message: `Error mapping function failed: ${error}`,
        context: { originalError: this.errorValue, error: String(error) },
        timestamp: Date.now(),
        cause: this.errorValue,
      } as QiError);
    }
  }

  recover(fn: (error: QiError) => T): Result<T> {
    try {
      return Result.success(fn(this.errorValue));
    } catch (error) {
      return Result.failure({
        category: "RuntimeError",
        message: `Recovery function failed: ${error}`,
        context: { originalError: this.errorValue, error: String(error) },
        timestamp: Date.now(),
        cause: this.errorValue,
      } as QiError);
    }
  }

  unwrapOr(defaultValue: T): T {
    return defaultValue;
  }

  isSuccess(): boolean {
    return false;
  }

  unwrap(): never {
    throw new Error(`Cannot unwrap failed Result: ${this.errorValue.message}`);
  }

  error(): QiError {
    return this.errorValue;
  }
}

/**
 * Utility functions for working with Results
 */
export namespace Result {
  /**
   * Sequence a list of Results into a Result of a list
   */
  export function sequence<T>(results: Result<T>[]): Result<T[]> {
    const values: T[] = [];
    for (const result of results) {
      if (result.isFailure()) {
        return result as Result<T[]>;
      }
      values.push(result.unwrap());
    }
    return Result.success(values);
  }

  /**
   * Traverse a list with a function that returns Results
   */
  export function traverse<T, U>(items: T[], fn: (item: T) => Result<U>): Result<U[]> {
    return sequence(items.map(fn));
  }

  /**
   * Convert all Results to values, collecting any errors
   */
  export function partition<T>(results: Result<T>[]): {
    successes: T[];
    failures: QiError[];
  } {
    const successes: T[] = [];
    const failures: QiError[] = [];

    for (const result of results) {
      if (result.isSuccess()) {
        successes.push(result.unwrap());
      } else {
        failures.push(result.error());
      }
    }

    return { successes, failures };
  }

  /**
   * Wrap a function that might throw into a Result
   */
  export function tryCatch<T>(fn: () => T): Result<T> {
    try {
      return Result.success(fn());
    } catch (error) {
      return Result.failure({
        category: "RuntimeError",
        message: `Function threw: ${error}`,
        context: { error: String(error) },
        timestamp: Date.now(),
      } as QiError);
    }
  }

  /**
   * Wrap an async function that might throw into a Result
   */
  export async function tryCatchAsync<T>(fn: () => Promise<T>): Promise<Result<T>> {
    try {
      const value = await fn();
      return Result.success(value);
    } catch (error) {
      return Result.failure({
        category: "RuntimeError",
        message: `Async function threw: ${error}`,
        context: { error: String(error) },
        timestamp: Date.now(),
      } as QiError);
    }
  }
}

File: haskell/QiCore/Base/Result.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE ScopedTypeVariables #-}

module QiCore.Base.Result
  ( Result (..)

    -- * Construction
  , success
  , failure
  , fromTryCatch
  , fromMaybe
  , fromEither

    -- * Transformation
  , mapResult
  , mapError
  , flatMap
  , andThen

    -- * Extraction
  , unwrap
  , unwrapOr
  , unwrapOrElse

    -- * Query
  , isSuccess
  , isFailure
  , getData
  , getError

    -- * Conditional operations
  , match
  , orElse
  , recover
  , retryOnFailure

    -- * Collection operations
  , sequenceResult
  , traverseResult
  , liftA2Result
  , filterSuccesses
  , filterFailures
  , partitionResults

    -- * Validation helpers
  , validateAll
  , validateFirst

    -- * Exception integration
  , QiErrorException (..)
  ) where

import Control.Applicative (Alternative (..), liftA2)
import Control.Exception (Exception, SomeException, throw, try)
import Data.Foldable (traverse_)
import Data.Text qualified as T
import Data.Typeable (Typeable)
import GHC.Generics (Generic)
import QiCore.Base.Error (QiError (..))
import QiCore.Base.Error qualified as Error

{- | Result type for operations that can succeed with a value or fail with a structured error

This type is similar to 'Either' but specialized for error handling with 'QiError'.
It provides:

* Clear success/failure semantics
* Rich error information via 'QiError'
* Monadic composition for error propagation
* Integration with exception handling
-}
data Result a = Success !a | Failure !QiError
  deriving stock (Eq, Show, Generic)

-- * Basic Construction

-- | Create a successful result
success :: a -> Result a
success = Success

-- | Create a failed result
failure :: QiError -> Result a
failure = Failure

-- | Convert a 'Maybe' to a 'Result' with a default error
fromMaybe :: QiError -> Maybe a -> Result a
fromMaybe err = \case
  Nothing -> Failure err
  Just a -> Success a

-- | Convert an 'Either' to a 'Result'
fromEither :: Either QiError a -> Result a
fromEither = \case
  Left err -> Failure err
  Right a -> Success a

{- | Safely execute an 'IO' action, converting exceptions to 'Result'

Any thrown exception is caught and converted to a 'QiError' with category 'UNKNOWN'.
-}
fromTryCatch :: IO a -> IO (Result a)
fromTryCatch action = do
  result <- try action
  case result of
    Right value -> pure $ Success value
    Left (ex :: SomeException) -> do
      -- Convert exception to QiError
      qiError <-
        Error.create
          "EXCEPTION_CAUGHT"
          (T.pack $ show ex)
          Error.UNKNOWN
          Nothing
          Nothing
      pure $ Failure qiError

-- * Transformation

-- | Transform the success value (Functor fmap)
mapResult :: (a -> b) -> Result a -> Result b
mapResult f = \case
  Success a -> Success (f a)
  Failure err -> Failure err

-- | Transform the error (contravariant map on error)
mapError :: (QiError -> QiError) -> Result a -> Result a
mapError f = \case
  Success a -> Success a
  Failure err -> Failure (f err)

-- | Chain operations that return Results (Monad bind)
flatMap :: (a -> Result b) -> Result a -> Result b
flatMap f = \case
  Success a -> f a
  Failure err -> Failure err

-- | Alias for 'flatMap' with arguments flipped (more intuitive for chaining)
andThen :: Result a -> (a -> Result b) -> Result b
andThen = flip flatMap

-- * Extraction

{- | Return the success value or throw the error as an exception

Use with caution - prefer 'unwrapOr' or 'match' for safer alternatives.
-}
unwrap :: Result a -> IO a
unwrap = \case
  Success a -> pure a
  Failure err -> throw (QiErrorException err)

-- | Return the success value or a default value
unwrapOr :: a -> Result a -> a
unwrapOr defaultValue = \case
  Success a -> a
  Failure _ -> defaultValue

-- | Return the success value or compute a default using the error
unwrapOrElse :: (QiError -> a) -> Result a -> a
unwrapOrElse f = \case
  Success a -> a
  Failure err -> f err

-- * Query

-- | Check if the result is a success
isSuccess :: Result a -> Bool
isSuccess = \case
  Success _ -> True
  Failure _ -> False

-- | Check if the result is a failure
isFailure :: Result a -> Bool
isFailure = not . isSuccess

-- | Extract the success value as 'Maybe'
getData :: Result a -> Maybe a
getData = \case
  Success a -> Just a
  Failure _ -> Nothing

-- | Extract the error as 'Maybe'
getError :: Result a -> Maybe QiError
getError = \case
  Success _ -> Nothing
  Failure err -> Just err

-- * Conditional Operations

-- | Pattern match on both success and failure cases
match :: (a -> b) -> (QiError -> b) -> Result a -> b
match onSuccess onError = \case
  Success a -> onSuccess a
  Failure err -> onError err

-- | Provide an alternative result if this one failed
orElse :: (QiError -> Result a) -> Result a -> Result a
orElse alternative = \case
  Success a -> Success a
  Failure err -> alternative err

-- | Recover from failure with a pure value
recover :: a -> Result a -> Result a
recover defaultValue = orElse (const $ Success defaultValue)

-- | Retry a computation on failure (useful for transient errors)
retryOnFailure :: Int -> IO (Result a) -> IO (Result a)
retryOnFailure maxAttempts action = go maxAttempts
 where
  go 0 = action -- Last attempt, return whatever we get
  go n = do
    result <- action
    case result of
      Success _ -> pure result
      Failure _ -> go (n - 1)

-- * Collection Operations

{- | Convert a list of 'Result's to a 'Result' of a list

Succeeds only if all results are successful.
-}
sequenceResult :: [Result a] -> Result [a]
sequenceResult = foldr go (Success [])
 where
  go (Success a) (Success as) = Success (a : as)
  go (Failure err) _ = Failure err
  go _ (Failure err) = Failure err

-- | Map a function over a list and collect the results
traverseResult :: (a -> Result b) -> [a] -> Result [b]
traverseResult f = sequenceResult . map f

-- | Lift a binary function to work on 'Result's
liftA2Result :: (a -> b -> c) -> Result a -> Result b -> Result c
liftA2Result f (Success a) (Success b) = Success (f a b)
liftA2Result _ (Failure err) _ = Failure err
liftA2Result _ _ (Failure err) = Failure err

-- | Extract all successful values from a list of results
filterSuccesses :: [Result a] -> [a]
filterSuccesses = foldr go []
 where
  go (Success a) acc = a : acc
  go (Failure _) acc = acc

-- | Extract all errors from a list of results
filterFailures :: [Result a] -> [QiError]
filterFailures = foldr go []
 where
  go (Success _) acc = acc
  go (Failure err) acc = err : acc

-- | Partition a list of results into successes and failures
partitionResults :: [Result a] -> ([a], [QiError])
partitionResults = foldr go ([], [])
 where
  go (Success a) (successes, failures) = (a : successes, failures)
  go (Failure err) (successes, failures) = (successes, err : failures)

-- * Validation Helpers

{- | Validate all items in a list, collecting all errors

Unlike 'traverseResult', this continues validation even after the first failure,
collecting all validation errors.
-}
validateAll :: (a -> Result b) -> [a] -> Result [b]
validateAll f items =
  let results = map f items
      (successes, failures) = partitionResults results
   in case failures of
        [] -> Success successes
        (firstError : _) -> Failure firstError -- Could combine errors here

-- | Validate items until the first success
validateFirst :: (a -> Result b) -> [a] -> Result b
validateFirst _ [] =
  Failure $
    QiError
      { Error.code = "NO_ITEMS"
      , Error.message = "Empty list for validation"
      , Error.category = Error.VALIDATION
      , Error.context = Nothing
      , Error.cause = Nothing
      , Error.timestamp = 0
      }
validateFirst f (x : xs) = case f x of
  Success b -> Success b
  Failure _ -> validateFirst f xs

-- * Exception Integration

-- | Exception wrapper for 'QiError' to enable throwing
newtype QiErrorException = QiErrorException QiError
  deriving stock (Show, Typeable)

instance Exception QiErrorException

-- * Type Class Instances

-- | Transform success values while preserving failures
instance Functor Result where
  fmap = mapResult

{- | Apply functions in 'Result' context

Accumulates the first error encountered.
-}
instance Applicative Result where
  pure = Success
  Success f <*> Success a = Success (f a)
  Success _ <*> Failure err = Failure err
  Failure err <*> _ = Failure err

{- | Chain computations that may fail

Short-circuits on the first failure.
-}
instance Monad Result where
  (>>=) = flip flatMap

-- | Provide alternative results on failure
instance Alternative Result where
  empty =
    Failure $
      QiError
        { Error.code = "EMPTY_ALTERNATIVE"
        , Error.message = "Empty Alternative result"
        , Error.category = Error.UNKNOWN
        , Error.context = Nothing
        , Error.cause = Nothing
        , Error.timestamp = 0
        }
  Success a <|> _ = Success a
  Failure _ <|> r = r
